Instagram/app/controllers/comments_controller.rb

```diff
@@ -0,0 +1,25 @@
+class CommentsController < ApplicationController
+
+  def create
+    image = Image.find(params[:image_id])
+    comment = image.comments.new(comment_params)
+    comment.save
+    if comment.errors.messages.empty?
+      redirect_to image
+    else
+      messages = ""
+      comment.errors.full_messages.each do |message|
+        messages += message + " "
+      end
+      flash[:notice] = messages
+      redirect_to image
+    end
+  end
```

Comment:
There are a few angles on this problem of having a lot of logic in the controller. The most crucial, and yet subtle, one is this — following rails conventions saves you work, contravening them makes work for you.

The core areas of incidental complexity here are:

1. You are managing the params and object creation yourselves, consider:

```ruby
class CommentsController < ApplicationController

  def create
    comment = Comment.new(comment_params)
    if comment.save
      redirect_to comment.image
    else
      messages = ""
      comment.errors.full_messages.each do |message|
        messages += message + " "
      end
      flash[:notice] = messages
      redirect_to comment.image
    end
  end

  private

  def comment_params
    params
      .require(:comment)
      .permit(:comment, :commenter)
      .merge({image_id: params[:image_id]})
  end

end
```

2. You have put the comment new form on the image page. This means it's harder to leverage rails form error handling (which autopopulates errors in a form for you).

There are good ways to make number 2 happen, but they require working with the rails architecture rather than against it. A good way to learn this is to start with a really standard CRUD app (the sort rails scaffolds generates — so `Comment#new` will be on a separate page) and then refactor towards the interface you want. After that, the right approach will become clearer.

You might not work in Rails again after this, but the approach applies to all opinionated frameworks — perfectly idiomatic at first, then learn how to express yourself in the idioms.